runtime/main/src/main/java/org/qbicc/runtime/main/ObjectModel.java

```diff
@@ -28,7 +28,97 @@
     /**
      * Get the java.lang.Class instance for the type ID.
      */
-    public static native Class<?> get_class_from_type_id(CNative.type_id typeId);
+    public static native Class<?> get_class_from_type_id(CNative.type_id typeId, Stdint.uint8_t dimensions);
+
+    /**
+     * Returns java.lang.Class instance representing array class of a given class
+     *
+     * @param componentClass
+     * @return instance of java.lang.Class
+     */
+    public static native Class<?> get_array_class_of(Class<?> componentClass);
+
+    /**
+     * Tries to atomically set the java.lang.Class#arrayClass field of a component class
+     * to a given array class
+     *
+     * @param componentClass
+     * @param arrayClass
+     * @return boolean true if atomic operation succeeds, false otherwise
+     */
+    public static native boolean set_array_class(Class<?> componentClass, Class<?> arrayClass);
+
+    /**
+     * Checks if the java.lang.Class instance represents reference array class.
+     * Reference array class have dimension greater than 0.
+     *
+     * @param cls
+     * @return boolean
+     */
+    public static boolean is_reference_array_class(Class<?> cls) {
+        CNative.type_id typeId = get_type_id_from_class(cls);
+        return is_reference_array(typeId);
+    }
+
+    /**
+     * Returns java.lang.Class instance representing the array class of a given component class
+     *
+     * @param componentClass
+     * @return instance of java.lang.Class
+     */
+    public static Class<?> get_or_create_array_class(Class<?> componentClass) {
+        Class<?> arrayClass = get_array_class_of(componentClass);
+        if (arrayClass == null) {
+            String className;
+            if (is_reference_array_class(componentClass)) {
+                className = "[" + componentClass.getName();
+                arrayClass = create_class(className, get_reference_array_typeid(), get_dimensions_from_class(componentClass));
```

Comment:
I have intentionally duplicated this statement in both `if` and `else` block.
Putting it at the end is resulting in some issue with LLVM code generation (LLVM IR looks fine I think) which results in incorrect `className` set for the newly created instance of j.l.Class when the code flows through `if` block.
I will try to create a simplified test case that reproduces this issue.